[ROCm] Workaround for Eigen::half bug related bug on ROCm platform

deven-amd opened the pull request:
copy-pasting the PR description from @ekuznetsov139 description for the same in ROCm fork

--------------

There appears to be an ABI mismatch between gcc7 and clang that, in case of functor::Pad<Eigen::half, 3>, results in the pad value being passed incorrectly (we see the correct value of 0 in pad_op.cc, which is compiled with gcc, and a random incorrect value in pad_op.h, which is compiled with clang).

This PR switches functor::Pad to taking all arguments by reference, which works around the problem.

It also adds float16 to tested data types for pad_op_test (the problem was not noticed for a long time because pad_op_test was not testing float16).

---------------

/cc @rmlarsen @cantonios ... this is the TF side workaround for the same issue/bug for which we filed this Eigen PR - https://gitlab.com/libeigen/eigen/-/merge_requests/422

----------------

/cc @cheshire @chsigg

deven-amd commented:
@googlebot I consent

gbaned commented:
@deven-amd Can you please sign CLA. Thanks!

cantonios commented:
Is this still needed with the recent Eigen change?

deven-amd commented:
@cantonios not technically needed if the TF Eigen pointer has been updated to pick the Eigen side fix.

But since this commit also includes a unit-test update that will fail on ROCm without the fix, I would prefer to get this PR merged.

cantonios commented:
> @cantonios not technically needed if the TF Eigen pointer has been updated to pick the Eigen side fix.
> 
> But since this commit also includes a unit-test update that will fail on ROCm without the fix, I would prefer to get this PR merged.

The Eigen pointer was updated last week - so I think we should be good there.  Can you remove the unnecessary changes and double-check it passes as expected?  We should merge the added test at the least.

deven-amd committed:
Adding float16 to list of types to test within pad_op_test.py:testFloatTypes

This test should fail for float16 (on ROCm, and possibly other GPU platforms too).

The fix for this is within Eigen code. See this PR for more details : https://gitlab.com/libeigen/eigen/-/merge_requests/422

deven-amd commented:
>  Can you remove the unnecessary changes and double-check it passes as expected? We should merge the added test at the least.

done.

@cantonios please re-approve

cantonios reviewed (approved):
LGTM
